core/src/main/java/com/linecorp/armeria/common/FlagsProvider.java

```diff
@@ -0,0 +1,923 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.nio.channels.ClosedChannelException;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Predicate;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLException;
+
+import com.github.benmanes.caffeine.cache.CaffeineSpec;
+
+import com.linecorp.armeria.client.ClientBuilder;
+import com.linecorp.armeria.client.ClientFactoryBuilder;
+import com.linecorp.armeria.client.DnsResolverGroupBuilder;
+import com.linecorp.armeria.client.retry.Backoff;
+import com.linecorp.armeria.client.retry.RetryingClient;
+import com.linecorp.armeria.client.retry.RetryingRpcClient;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.common.util.Sampler;
+import com.linecorp.armeria.common.util.SystemInfo;
+import com.linecorp.armeria.common.util.TransportType;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.TransientService;
+import com.linecorp.armeria.server.TransientServiceOption;
+import com.linecorp.armeria.server.file.FileService;
+import com.linecorp.armeria.server.file.FileServiceBuilder;
+import com.linecorp.armeria.server.file.HttpFile;
+
+import io.netty.channel.ChannelOption;
+import io.netty.channel.EventLoopGroup;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.resolver.DefaultAddressResolverGroup;
+import io.netty.resolver.dns.DnsNameResolverTimeoutException;
+
+/**
+ * A Java SPI (Service Provider Interface) for the {@link Flags} values. Returning null to indicates that this
+ * FlagsProvider doesn't provide the flag.
+ *
+ * <p>Two {@link FlagsProvider}s are provided by default.
+ * <ul>
+ *     <li>The system property {@link FlagsProvider} which provides value from JVM option</li>
+ *     <li>The default {@link FlagsProvider} which provides the default values when flag isn't provides or
+ *         fail</li>
+ * </ul>
+ *
+ */
+@UnstableApi
+@FunctionalInterface
+public interface FlagsProvider {
```

Comment:
How about adding `FlagsProvider.name()` which returns a human-readable name of the provider?
```
public interface FlagsProvider {
  default String name() {
    return getClass().getSimpleName();
  }
}

class DefaultFlagsProvider ... {
  @Override
  public String name() {
    return "default";
  }
  ...
}

class SystemPropertyFlagsProvider ... {
  @Override
  public String name() {
    return "sysprops";
  }
  ...
}
```

Then we could use this name in the log messages in `Flags` to get prettier output.

Reply:
That sounds good to me 👍. I'll implement that. Thanks